### ISSUE TITLE
stringifyArgs throws error on "non-scalar" scalar type

### ISSUE THREAD

**Issue by awinograd** [typed-graphql-builder/typed-graphql-builder#52]
```gql
"""
The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSONObject

input MyInput {
  content: JSONObject
}

type Mutation {
    createObject(input: MyInput!): Boolean!
}
```

The above schema will throw the following error when trying to build a mutation with `content` as an argument since JSONObject does not exist in `$InputTypes`

>                 throw new Error(`Argument type for ${key} not found`)

**Comment by spion** [comment]
Are you using https://typed-graphql-builder.spion.dev/docs/custom-scalars or something else? Could we assume that `JSON.stringify` will work for all custom scalar serialization or do we need customization?

**Comment by awinograd** [comment]
@spion yes I'm using custom scalars importing from a separate file. I tried poking around the source myself to see if I could come up with a solution. Unfortunately JSON.stringify by itself does not work. The embedded object type in graphql doesn't follow JSON. For example, keys don't have quotes.

My workaround for now has been to use a `$` variable and pass the json object in at request time. That works as expected

**Comment by spion** [comment]
Using variables seems to be the recommended way as far as I can tell. There are custom serializers and deserializers in Apollo but I'm not entirely clear we need that. A better error message might help...

**Comment by spion** [comment]
I wrote a first pass PR in 56 that at least produces a helpful error message for now.

**Comment by awinograd** [comment]
@spion I think I've stumbled across another issue using JsonObject as a scalar. I'm not sure how this was previously working for me in some instances but in some usages `$` is failing to infer the variable type for inputs with JsonObject.

I think the core issue is the use of `$Atomic` in `VariabledInput`

```ts
// the array wrapper prevents distributive conditional types
// https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types
type VariabledInput<T> = [T] extends [$Atomic | null | undefined]
  ? Variable<T, any> | T
  : T extends ReadonlyArray<infer I>
  ? Variable<T, any> | T | ArrayInput<I>
  : T extends Record<string, any>
  ? Variable<T, any> | { [K in keyof T]: VariabledInput<T[K]> } | T
  : never
```

Since `$Atomic = JSONObject | number | boolean | ... ` includes `JSONObject` the ternary operator for VariabledInput seems to always be returning `true` so we're never getting to the Record case below.


Is there a reason behind checking against `$Atomic` first instead of trying to match against arrays and records first beforehand?

**Comment by spion** [comment]
That would be the point of atomic (including scalar) types - they are not divisible and GraphQL doesn't understand how to divide them. They have to be sent to the server in one go, so there cannot exist a variable for a subset of an atomic type.

**Comment by awinograd** [comment]
I agree with that understanding from a GraphQL perspective. However, in this case for typescript, the GraphQL scalar is not necessarily a Typescript scalar.

`$Atomic` assumes that a GraphQL scalar is always a typescript scalar

**Comment by spion** [comment]
Regardless, in typescript, we can't do

```
someQuery({myScalar: {inner: $('variable')}}
```

because that cannot translate to valid graphql. That's why recursion stops at the scalar

**Comment by awinograd** [comment]
I dont understand why that wouldn't be valid GQL. Though maybe I'm not following your property names in the above example properly

```graphql
query MyQuery($variable: JSONObject) {
  someQuery(myScalar: { scalarInnerProperty: $variable } }) {
    foo
  }
}
```

```ts
const myJsonObject = { foo: 'bar' }
client.query(MyQueryDocument, { scalarInnerProperty: myJsonObject })
```

**Comment by spion** [comment]
```
someQuery(myScalar: { scalarInnerProperty: $variable } }) 
```

If the schema for the input type says `myScalar: JSONObject`, and the schema says `scalar JSONObject` then from the perspective of GraphQL `scalarInnerProperty` does not exist.

**Comment by awinograd** [comment]
Ok, I understand how I misinterpreted your example. I agree with that but I dont think that's quite what I'm trying to get at with the use of the JSONScalar object. It is valid for the scalar to exist inside a parent object. 

```gql
type SomeArgs {
  name: String!
  content: JsonObject!
}

type Query {
    someQuery(args: SomeArgs!): Boolean!
}

query MyQuery($variable: JSONObject) {
  someQuery(args: { name: "Baz", content: $variable } }) {
    foo
  }
}
```

It should be valid to build `MyQuery` with `typed-graphql-builder` as follows:

```ts
const MyQueryDocument = query(q => [
  q.someQuery({ name: "Baz", content: $('content' )})
])

client.query(MyQueryDocument, { content: { foo: 'bar' } });

```

**Comment by awinograd** [comment]
@spion The following definition for `VariabledInput` seems to work for me and the typed-graphql-builder test suite passes. I will try to get a PR open with a regression test for the above example. Does that sound good to you?

```ts
type VariabledInput<T> = [T] extends [Exclude<$Atomic, Record<string, any>> | null | undefined]
  ? Variable<T, any> | T
  : T extends ReadonlyArray<infer I>
  ? Variable<T, any> | T | ArrayInput<I>
  : T extends Record<string, any>
  ? Variable<T, any> | { [K in keyof T]: VariabledInput<T[K]> } | T
  : never
```

**Comment by spion** [comment]
I see. Your example should work indeed.

I would like to understand why this isn't working first, as that particular generic has very tricky / edge-casy behavior.

**Comment by awinograd** [comment]
@spion I think it's not working due to the following:

`SomeArgs` is an object type.

`VariabledInput<SomeArgs>` will always return `Variable<T, any> | T` if JSON is provided as a scalar since `object extends JSON` is true.

So, if you try to build a query with inline args `q.someQuery({ name: '', content: $('content') })` the type expected by someQuery is only a variable for the entire object or the inline object. We never get the option to specify some keys manually and others as variables b/c we never reach `{ [K in keyof T]: VariabledInput<T[K]> }`

**Comment by awinograd** [comment]
You're welcome to use all/any of my PR or close it out with your preferred implementation as you'd like. Thanks as always for this library!

**Comment by spion** [comment]
I see, the scalar type is defined broad enough to include all other input types

The problem with this change is that it will also allow invalid queries i.e. using a variable in a scalar.

Is there any chance to be more specific with the type? If I can't be specific I usually use `unknown` + a validator, but if I know the schema I go with a more specific type

**Comment by awinograd** [comment]
@spion I don't believe my change in the PR allows for using a variable in a scalar or other invalid queries (note that the PR is different from my comment above). 

If i try to add a variable inside a JSONObject scalar I get:

> Type instantiation is excessively deep and possibly infinite. (tsserver 2589)

If a custom scalar has a defined record type (e.g. `scalar SpecificShape = { foo: string }`) I think `VariabledInput` will likely allow the variable-in-scalar "bug" you talked about. However, I think the current implementation just wouldn't handle it at all, so it doesn't have the same "bug" b/c the typechecking fails before then. 

I guess the difference is the current implementation fails fast while my PR typechecks even though the code is invalid... 

It's not immediately clear to me how to workaround that issue, but I'll keep thinking about it and see if I come up with any ideas

**Comment by spion** [comment]
With some branding (hidden from the users) I think this is doable: https://github.com/typed-graphql-builder/typed-graphql-builder/pull/58

**Comment by spion** [comment]
Merged with #58 and shold be available in a major version bump
